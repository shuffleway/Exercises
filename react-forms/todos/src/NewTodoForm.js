import React, { useState } from "react";
import './TodoList.css';

const NewTodoForm = ({ addTodo }) => {
  const initialState = { item: "" };
  const [formData, setFormData] = useState(initialState);
  const [idCounter, setIdCounter] = useState(1);
  const [errors, setErrors] = useState({});

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value
    }));
  };

  const validate = () => {
    const newErrors = {};
    if (!formData.item) {
      newErrors.item = "Todo item is required.";
    }
    return newErrors;
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const newErrors = validate();
    if (Object.keys(newErrors).length > 0) {
      setErrors(newErrors);
    } else {
      addTodo({ ...formData, id: idCounter });
      setIdCounter(idCounter + 1);
      setFormData(initialState);
      setErrors({});
    }
  };

  return (
    <div className="new-todo-form">
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="item">Add Todo: </label>
          <input 
            id="item"
            type="text"
            name="item"
            value={formData.item}
            onChange={handleChange}
          />
          <button>Add Todo</button>
          {errors.item && <div className="error">{errors.item}</div>}
        </div>  
      </form>
    </div>
  );
};

export default NewTodoForm;
