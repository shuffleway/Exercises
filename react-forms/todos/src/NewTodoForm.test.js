import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";

it("renders without crashing", function() {
  render(<NewTodoForm />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<NewTodoForm />);
  expect(asFragment()).toMatchSnapshot();
});

// it("runs the create function on form submit", function() {
//   const createMock = jest.fn();
//   const { getByText } = render(<NewTodoForm addTodo={createMock} />);
//   const createButton = getByText("Add Todo!");
//   fireEvent.click(createButton);
//   expect(createMock).toHaveBeenCalled();
// });



it("runs the create function on form submit", function() {
  const addTodoMock = jest.fn();
  const { getByText, getByLabelText } = render(<NewTodoForm addTodo={addTodoMock} />);
  
  // Fill the input field
  const input = getByLabelText("Add Todo:");
  fireEvent.change(input, { target: { value: "Test Todo" } });
  
  // Submit the form
  const createButton = getByText("Add Todo");
  fireEvent.submit(createButton.closest('form'));
  
  expect(addTodoMock).toHaveBeenCalled();
});
