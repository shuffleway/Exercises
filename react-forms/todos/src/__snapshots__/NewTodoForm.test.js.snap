// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="new-todo-form"
  >
    <form>
      <div>
        <label
          for="item"
        >
          Add Todo: 
        </label>
        <input
          id="item"
          name="item"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
